### PR TITLE
Validate syntax for incoming notifications (#736)

### PR DESCRIPTION
**Pull request by achave11**

### CODE REVIEW

File: lambdas/indexer/app.py [hunk 1]
```diff
@@ -98,12 +99,38 @@ def post_notification():
         raise chalice.ChaliceViewError(test_mode_error)
 
 
+def validate_request_syntax(notification):
+    match_ = notification['match']
```
Comment (hannes-ucsc) [inline lambdas/indexer/app.py]: If `match` is absent we should return a 400. This currently returns 500.

File: lambdas/indexer/app.py [hunk 2]
```diff
@@ -79,6 +79,7 @@ def post_notification():
     notification = app.current_request.json_body
     log.info("Received notification %r", notification)
     params = app.current_request.query_params
+    validate_request_syntax(notification)
```
Comment (hannes-ucsc) [inline lambdas/indexer/app.py]: Move one line up.

File: test/indexer/test_hca_indexer.py [hunk 3]
```diff
@@ -551,6 +560,99 @@ def test_organoid_priority(self):
...
+    def test_post_notification_endpoint(self):
+        self._create_mock_notify_queue()
+        body = {
+            'match': {
+                'bundle_uuid': 'bb2365b9-5a5b-436f-92e3-4fc6d86a9efd',
+                'bundle_version': '2018-03-28T13:55:26.044Z'
+            }
+        }
+        for endpoint in ['/', '/delete']:
+            with self.subTest(f'Test correct format notification to "{endpoint}" endpoint'):
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py]:
```suggestion
            with self.subTest(endpoint=endpoint):
```

File: test/indexer/test_hca_indexer.py [hunk 4]
```diff
@@ -551,6 +560,99 @@ def test_organoid_priority(self):
...
+                response = self._test(body, endpoint)
+                self.assertEqual(200 if endpoint == '/' else 202, response.status_code)
+                self.assertEqual({'status': 'done'}, response.json()) if endpoint == '/' else None
+
+    @mock_sts
+    @mock_sqs
+    def test_post_notification_with_none(self):
+        for endpoint in ['/', '/delete']:
+            with self.subTest("Missing bundle_uuid"):
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py]:
```suggestion
            with self.subTest("Missing bundle_uuid", endpoint=endpoint):
```

File: test/indexer/test_hca_indexer.py [hunk 5]
```diff
@@ -551,6 +560,99 @@ def test_organoid_priority(self):
...
+                    'match': {
+                        'bundle_uuid': 'bb2365b9-5a5b-436f-92e3-4fc6d86a9efd',
+                        'bundle_version': None
+                    }
+                }, endpoint)
+                self.assertEqual(400, response.status_code)
+
+    @mock_sts
+    @mock_sqs
+    def test_post_notification_with_malformed_body(self):
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py]: I would merge test_post_notification_with_none and this method.

File: test/indexer/test_hca_indexer.py [hunk 6]
```diff
@@ -551,6 +560,99 @@ def test_organoid_priority(self):
...
+        }
+        for endpoint in ['/', '/delete']:
+            with self.subTest(f'Test correct format notification to "{endpoint}" endpoint'):
+                response = self._test(body, endpoint)
+                self.assertEqual(200 if endpoint == '/' else 202, response.status_code)
+                self.assertEqual({'status': 'done'}, response.json()) if endpoint == '/' else None
+
+    @mock_sts
+    @mock_sqs
+    def test_post_notification_with_none(self):
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py]: This doesn't just test none.

File: test/indexer/test_hca_indexer.py (hunk 4 above)
Comment (hannes-ucsc) [reply test/indexer/test_hca_indexer.py]: Or, you can nest subTests

File: test/indexer/test_hca_indexer.py [hunk 7]
```diff
@@ -551,6 +560,99 @@ def test_organoid_priority(self):
...
+        for endpoint in ['/', '/delete']:
+            with self.subTest("Missing bundle_uuid"):
+                response = self._test({
+                    'match': {
+                        'bundle_version': '2018-03-28T13:55:26.044Z'
+                    }
+                }, endpoint)
+                self.assertEqual(400, response.status_code)
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py]: This assertion is repeated multiple times. Consider refactoring.

File: test/indexer/test_hca_indexer.py [hunk 8]
```diff
@@ -551,6 +560,93 @@ def test_organoid_priority(self):
...
+                response = self._test(body, endpoint)
+                self.assertEqual(202, response.status_code)
+                self.assertEqual('', response.text)
+
+    @mock_sts
+    @mock_sqs
+    def test_invalid_notifications(self):
+        for endpoint in ['/', '/delete']:
+            with self.subTest(endpoint=endpoint):
+                bodys = {
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py]:
```suggestion
                bodies = {
```

Comment (hannes-ucsc) [reply test/indexer/test_hca_indexer.py]: Can we pull this dict literal to the beginning of the method, outside of the loop and any subTest contexts?

File: lambdas/indexer/app.py [hunk 9]
```diff
@@ -91,19 +92,49 @@ def post_notification():
...
+        return chalice.app.Response(body='', status_code=http.HTTPStatus.ACCEPTED)
     else:
         test_mode_error = f'Ignored notification {notification}. This indexer is currently in TEST MODE.'
         log.error(test_mode_error)
         raise chalice.ChaliceViewError(test_mode_error)
 
 
+def validate_request_syntax(notification):
+    try:
+        match_ = notification['match']
```
Comment (hannes-ucsc) [inline lambdas/indexer/app.py]: Why is there an underscore at the end of `match_`?

```suggestion
        match_ = notification['match']
```

Comment (achave11) [reply lambdas/indexer/app.py]: PyCharm suggestion. 😅

**Pull request merged**